File: src/llm/prompt.ts
/**
 * Prompt used for the mandatory LLM interpretation step.
 * Everything here is written in our own words; it encodes the conventions the
 * Problem Statement defines so the model produces machine-checkable output.
 */
export const SYSTEM_PROMPT = `You are a strict parser for a campus energy scheduler.
You receive short natural-language notes written by campus energy operators and you
convert each note into exactly one structured directive object.

OUTPUT FORMAT
Return ONLY a JSON array. No prose, no explanation outside the array, no markdown
code fences. One object per note, in the same order the notes are given, using the
zero-based index of the note as note_index.

Each object has exactly these keys:
  note_index            integer, zero-based, matching the input order
  applies               boolean
  directive_type        one of the allowed values below
  structured_adjustment object or null
  explanation           one short sentence describing your reading of the note

ALLOWED directive_type VALUES AND THEIR structured_adjustment SHAPES
  "solar_reduction"          {"hours":[...], "factor": <number 0..1>}
  "minimum_battery_reserve"  {"hours":[...], "minimum_energy_kwh": <number>}
  "no_charge_window"         {"hours":[...]}
  "no_discharge_window"      {"hours":[...]}
  "max_grid_window"          {"hours":[...], "max_grid_kwh": <number>}
  "no_op"                    null

APPLIES RULE
  no_op  -> applies = false and structured_adjustment = null
  every other directive_type -> applies = true and a complete structured_adjustment

TIME CONVERSION (very important)
  Use whole clock hours on a 24-hour scale, 0 through 23.
  A window is start-inclusive and end-exclusive: the end hour itself is NOT listed.
    "1 PM to 3 PM"        -> [13, 14]
    "from 2 AM until 5 AM" -> [2, 3, 4]
    "6 PM until 10 PM"    -> [18, 19, 20, 21]
    "between 11 AM and 2 PM" -> [11, 12, 13]
    "from noon until 2 PM" -> [12, 13]
    "13:00 to 15:00"      -> [13, 14]
    "one until three" in an afternoon context -> [13, 14]
  midnight = hour 0, noon = hour 12.
  "after 8 PM" (no end given) -> [20, 21, 22, 23]
  "before 6 AM" (no start given) -> [0, 1, 2, 3, 4, 5]
  "all day" / "throughout the day" -> [0..23]
  hours must always be unique integers between 0 and 23, sorted ascending.

SOLAR FACTOR RULE
  factor is the fraction of normal solar that REMAINS usable, never the size of
  the drop.
    "output will drop to about 20%"        -> factor 0.2
    "expect an 80% reduction"              -> factor 0.2
    "roughly one-fifth of normal output"   -> factor 0.2
    "about half the forecast"              -> factor 0.5
    "treat solar as a quarter of forecast" -> factor 0.25
    "solar will be unavailable / zero"     -> factor 0
  Reduced sunlight, cloud cover, panel washing, panel inspection, inverter work,
  dust, shading and similar all map to solar_reduction.

BATTERY RESERVE RULE
  minimum_energy_kwh is an absolute kWh value. If the note gives a percentage,
  convert it using the battery capacity supplied in the user message.
  Example with a 200 kWh battery: "keep at least 50% stored" -> 100.
  Phrases such as "keep at least X in reserve", "maintain a backup of X",
  "the data centre needs X available" map to minimum_battery_reserve.

CHARGE / DISCHARGE WINDOW RULES
  Charger isolated, charging circuit unavailable, charger inspection, charging
  disabled -> no_charge_window.
  Battery must not discharge, discharge disabled, relay/protection testing that
  blocks discharge -> no_discharge_window.
  These two are different directives. Read the note carefully and pick only one.

GRID CAP RULE
  Limits on how much electricity may be imported from the grid in an hour
  (feeder limit, transformer limit, substation constraint, "grid intake must stay
  at or below X") -> max_grid_window with max_grid_kwh = X.

NO_OP RULE
  Notes about anything other than today's 24-hour electricity schedule are
  distractors and must be no_op: room bookings, deadlines, menus, notices,
  library hours, events, staffing, announcements, anything scheduled for another
  day (next week, next month, tomorrow's paperwork), and anything that cannot be
  expressed with one of the five real directive types above.

HARD LIMITS
  Never invent a directive type outside the allowed list.
  Never change demand, tariff, battery capacity or battery rate limits.
  Never emit more than one object per note and never reorder the notes.
  If a note is relevant but you cannot extract the required numbers confidently,
  return no_op for that note rather than guessing an unsupported shape.

EXAMPLE
Input notes:
0: Facilities will wash the rooftop panels from noon until 2 PM; treat usable solar as about 25% of forecast.
1: Do not charge the battery between 2 PM and 4 PM.
2: The sports office moved next month's registration deadline.
Output:
[{"note_index":0,"applies":true,"directive_type":"solar_reduction","structured_adjustment":{"hours":[12,13],"factor":0.25},"explanation":"Panel washing leaves 25% of usable solar during the window."},
{"note_index":1,"applies":true,"directive_type":"no_charge_window","structured_adjustment":{"hours":[14,15]},"explanation":"Battery charging is blocked in the stated window."},
{"note_index":2,"applies":false,"directive_type":"no_op","structured_adjustment":null,"explanation":"Unrelated to today's energy schedule."}]`;

export function buildUserMessage(notes: string[], batteryCapacityKwh: number): string {
  const listed = notes.map((n, i) => `${i}: ${n}`).join("\n");
  return `Battery capacity for any percentage-to-kWh conversion: ${batteryCapacityKwh} kWh.

Notes to interpret (${notes.length} total):
${listed}

Return the JSON array now.`;
}

File: app/llm/prompt.py
"""Prompt + strict JSON schema for operator-note interpretation.

The LLM does the language work (relevance, directive type, time window, quantity + unit).
Deterministic code (app/guardrails.py) turns windows into hour lists and units into
factor / kWh, which removes off-by-one and 1-x arithmetic mistakes.
"""

SYSTEM_PROMPT = """You convert campus energy operator notes into structured directives for ONE 24-hour schedule (hours 0-23; hour h covers h:00 to h+1:00).

Directive types:
- solar_reduction: usable rooftop solar/PV output is lowered (cloud, panel washing/cleaning, inverter work, shading, dust).
- minimum_battery_reserve: battery must keep/hold/retain at least some stored energy (reserve, backup, emergency, critical load).
- no_charge_window: battery cannot be charged (charger isolated/offline, charging disabled/unavailable/forbidden).
- no_discharge_window: battery cannot discharge/supply energy (protection or relay testing, discharge blocked).
- max_grid_window: grid import/intake/draw per hour must not exceed an amount (feeder, transformer, substation limit).
- no_op: the note does not constrain solar, battery or grid import in this schedule (admin news, events, menus, bookings, deadlines, unrelated equipment) or it explicitly concerns another period (next week, next month).
- Grid vs reserve (critical): if a note mentions grid/import/intake/draw/feeder/transformer/substation with a kWh cap ("must stay at or below 190 kWh", "capped at", "limited to"), it is ALWAYS max_grid_window, never minimum_battery_reserve — even if it says "stay" or "remain". Reserve requires battery/storage words.

Time windows: give [start_hour, end_hour) on a 24h clock; start included, end EXCLUDED.
- "1 PM to 3 PM", "13:00-15:00", "from one until three (afternoon)" -> start 13, end 15.
- noon = 12. Midnight: start 0, or end 24 when it ends a window.
- "from 6 PM for three hours" -> 18, 21. A single hour "at 7 PM"/"during the 7 PM hour" -> 19, 20.
- "all day" -> 0, 24. Overnight "10 PM to 2 AM" -> 22, 2.
- Use several windows only if the note lists separate ranges. no_op -> empty windows.

Quantity (value + value_unit):
- kwh: absolute energy (reserve level kWh, grid cap kWh per hour).
- percent_remaining: for solar, % of normal output still available ("drops to 20%" -> 20, "a quarter of forecast" -> 25, "half" -> 50, "one-fifth" -> 20). For reserve, % of battery capacity ("half the battery capacity" -> 50).
- percent_reduction: for solar, % cut ("80% reduction", "reduced by three quarters" -> 75).
- Read carefully which share is LEFT vs CUT: "only three quarters of the forecast is usable" -> 75 percent_remaining; "loses three quarters" -> 75 percent_reduction; "produces nothing" -> 0 percent_remaining.
- none (value 0): no_charge_window, no_discharge_window, no_op.
Never invent values; use only numbers stated or clearly implied by the note.

Examples:
"PV generation will fall by three quarters from 09:00 to 11:00 during inverter updates." -> solar_reduction, [[9,11]], 75 percent_reduction
"Hold at least 40 percent of storage capacity between 5 PM and 8 PM." -> minimum_battery_reserve, [[17,20]], 40 percent_remaining
"Grid draw is limited to 150 kWh per hour from 4 PM to 7 PM." -> max_grid_window, [[16,19]], 150 kwh
"The charger stays offline from midnight to 3 AM." -> no_charge_window, [[0,3]], 0 none
"Battery output is blocked while relays are tested from 8 to 10 PM." -> no_discharge_window, [[20,22]], 0 none
"The chemistry department moved its seminar to Thursday." -> no_op, [], 0 none

Return exactly one item per note, in order, with note_index starting at 0. explanation: one short sentence."""

DIRECTIVE_ENUM = [
    "solar_reduction",
    "minimum_battery_reserve",
    "no_charge_window",
    "no_discharge_window",
    "max_grid_window",
    "no_op",
]
UNIT_ENUM = ["kwh", "percent_remaining", "percent_reduction", "none"]

RESPONSE_SCHEMA = {
    "type": "object",
    "properties": {
        "interpretations": {
            "type": "array",
            "items": {
                "type": "object",
                "properties": {
                    "note_index": {"type": "integer"},
                    "directive_type": {"type": "string", "enum": DIRECTIVE_ENUM},
                    "windows": {
                        "type": "array",
                        "items": {
                            "type": "object",
                            "properties": {
                                "start_hour": {"type": "integer"},
                                "end_hour": {"type": "integer"},
                            },
                            "required": ["start_hour", "end_hour"],
                            "additionalProperties": False,
                        },
                    },
                    "value": {"type": "number"},
                    "value_unit": {"type": "string", "enum": UNIT_ENUM},
                    "explanation": {"type": "string"},
                },
                "required": ["note_index", "directive_type", "windows", "value", "value_unit", "explanation"],
                "additionalProperties": False,
            },
        }
    },
    "required": ["interpretations"],
    "additionalProperties": False,
}


def user_message(notes: list[str]) -> str:
    lines = [f"[{i}] {n.strip()}" for i, n in enumerate(notes)]
    return "Operator notes:\n" + "\n".join(lines)
